designs/0005-static-matrices.md

```diff
@@ -0,0 +1,240 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a generalization of the reverse mode autodiff type in Stan Math,
+`var_value<T>`, where `T` is the underlying storage type of the autodiff variable.
+For `T == double`, this type is equivalent to the current Stan Math `var` type.
+
+The usefulness of this type comes from setting `T` to more general data structures,
+in particular the Eigen Matrix types. Instead of dealing with matrices of vars,
+this allows operations in the Math library to be written in terms of vars that are
+themselves matrices.
+
+This makes working with matrix variables in Stan autodiff much faster. This design
+is fully backwards compatible with the curren Stan autodiff types and those remain.
+
+For the rest of this proposal, "autodiff" means "reverse mode autodiff".
+
+With optimizations from the compiler the conditions for a `matrix` to be a
+constant matrix can be detected and applied to Stan programs automatically
+for users.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, an `N x M` matrix in Stan is represented as an Eigen matrix holding
+a pointer to an underlying array of autodiff objects, each of those holding
+a pointer to the underlying autodiff object implementation.
+
+The basic Eigen type used by Stan is `Eigen::Matrix<var, R, C>`. For brevity,
+this will be referred to as a "matvar" type (a matrix of vars). There are no
+major differences between a matrix of vars or a vector of vars or a row vector
+of vars.
+
+This proposal will enable an `N x M` Stan matrix to be defined in terms of one
+autodiff object with two separate `N x M` containers, one for its values and
+one for its adjoints.
+
+This basic example of this type is `var_value<Eigen::MatrixXd>`. For brevity,
+these will be referred to as "varmat" types (a var that is a matrix). Again,
+there are no major differences when the matrix is replaced by a vector or a row vector.
+
+The motivation for introducing the varmat type is speed. These types speed up
+computations in a number of ways:
+
+1. The values in a varmat are stored separately from the `adjoints` and can
+be accessed independently. In a matvar the values and adjoints are interleaved.
+With how memory loads work on modern cpus, that means reading the values requires
+reading the values and the adjoints.
+
+2. To read all `N x M` values or adjoints of a varmat requires one pointer
+dereference and `N x M` double reads. Reading a similar matvar requires
+`N x M` pointer dereferences and `N x M` reads.
+
+3. Memory is linear in varmat, and in the best case interleaved in matvar
+(though because every element is a pointer, in worse case matvar memory accesses
+are entirely random).
```

Comment:
Another advantage of linear memory layout is that vectorized memory access instructions can be used to access it. This is a precondition for Eigen doing any vectorization.